### PR TITLE
Add Bilkent Cxtmenu

### PR DESCRIPTION
**Pull request by mj3cheun**
## About

Adds the bilkent context menu to dash-cytoscape using an API and implementation which will allow the toronto context menu to be added as well using the same format. Replaces existing context menu PRs.

- Adds a `cxtmenu` property to provide configuration for the context menu
- Adds a `cxtmenuData` property to receive callbacks from the context menu 

Note: please let me know what dash-cytoscape version an update including this addition along with the map addition should hold. I will update the change log with it once I know.

 

## Pre-Merge checklist
- [x] The project was correctly built with `npm run build:all`.
- [x] If there was any conflict, it was solved correctly.
- [ ] All changes were documented in CHANGELOG.md.
- [ ] All tests on CircleCI have passed.
- [ ] All Percy visual changes have been approved.
- [ ] Two people have :dancer:'d the pull request. You can be one of these people if you are a Dash Cytoscape core contributor.

## Reference Issues

Closes #[issue number]

## Other comments

### CODE REVIEW

File: src/lib/components/Cytoscape.react.js
Comment (alexcjohnson) [inline src/lib/components/Cytoscape.react.js]: unintentional change?

File: src/lib/components/Cytoscape.react.js
Comment (alexcjohnson) [inline src/lib/components/Cytoscape.react.js]:
```suggestion
     * Displays a context menu on right click. Requires extra layouts loaded.
     * Accepts a list of dictionaries, each of which describes
     * a context menu option. Options are rendered in the order presented.
```

**Review by alexcjohnson — APPROVED** [review]
Looks good, just two minor comments. 💃

File: src/lib/components/Cytoscape.react.js
Comment (mj3cheun) [reply src/lib/components/Cytoscape.react.js]: Oops yes unintentional change. Reverted.

File: src/lib/components/Cytoscape.react.js
Comment (mj3cheun) [reply src/lib/components/Cytoscape.react.js]: Sounds good

File: dash_cytoscape/Cytoscape.py
Comment (mtwichan) [inline dash_cytoscape/Cytoscape.py:70]: Typo.

`Dictionary returned when a context menu option is selected`

**Review by mtwichan — COMMENTED** [review]
I'm running into an error installing the library.

```
C:\Users\Dell\Desktop\Code\Zyphr\dash-cytoscape\dist>pip install dash_cytoscape-0.3.0.tar.gz
Processing c:\users\dell\desktop\code\zyphr\dash-cytoscape\dist\dash_cytoscape-0.3.0.tar.gz
    ERROR: Command errored out with exit status 1:
     command: 'c:\users\dell\appdata\local\programs\python\python39\python.exe' -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-req-build-vjwdxxaw\\setup.py'"'"'; __file__='"'"'C:\\Users\\Dell\\AppData\\Local\\Temp\\pip-req-build-vjwdxxaw\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base 'C:\Users\Dell\AppData\Local\Temp\pip-pip-egg-info-gleh_o9b'
         cwd: C:\Users\Dell\AppData\Local\Temp\pip-req-build-vjwdxxaw\
    Complete output (5 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\Dell\AppData\Local\Temp\pip-req-build-vjwdxxaw\setup.py", line 7, in <module>
        with open('package.json') as f:
    FileNotFoundError: [Errno 2] No such file or directory: 'package.json'
    ----------------------------------------
WARNING: Discarding file:///C:/Users/Dell/Desktop/Code/Zyphr/dash-cytoscape/dist/dash_cytoscape-0.3.0.tar.gz. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

I'm building the tarball for the library using `python setup.py sdist` and then attempting to install the library by `pip install dash_cytoscape-0.3.0.tar.gz`. Am I doing something wrong?

@maxkfranz @mj3cheun

File: src/lib/components/Cytoscape.react.js
Comment (alexcjohnson) [inline src/lib/components/Cytoscape.react.js]:
```suggestion
     * Dictionary returned when a context menu option is selected. Read-only.
```
The source of the typo @mtwichan pointed out

**Comment by alexcjohnson** [comment]
> I'm running into an error installing the library.

Hmm yeah, there's a bunch of issues here. I can install v0.3.0 from PyPI but I can't import it in Python:
```
>>> import dash_cytoscape as cy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/alex/plotly/dash-cytoscape/dash_cytoscape/__init__.py", line 23, in <module>
    with open(_filepath) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/alex/plotly/dash-cytoscape/dash_cytoscape/package-info.json'
```
Presumably I can install it because it has a wheel (which is great! keep building wheels along with the sdist tarballs!). If I try to build as you did, I get a similar problem. Has this really been broken since 0.3.0 was released in May? Specifically I think we need to revert these two changes:
- [looking for package.json in the root instead of inside the package dir](https://github.com/plotly/dash-cytoscape/commit/1812577ca2013821fe403b84ed74a1d76b4f4e65#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L7)
- [calling it package-info.json instead of package.json when copied in there](https://github.com/plotly/dash-cytoscape/commit/1812577ca2013821fe403b84ed74a1d76b4f4e65#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L7) - ie keep the rest of the change to `build:backends`, just remove the `-p package-info.json` so it's still called `package.json`, because that's what's in `MANIFEST.in` and it's what's referenced in [`__init__.py`](https://github.com/plotly/dash-cytoscape/blob/3799bbdfbc2b631a7cea98de67a7c5eeacc8ee78/dash_cytoscape/__init__.py#L22)

**Comment by mtwichan** [comment]
Hi folks,
 
What was the reasoning for naming the context menu property `cxtmenuData`/`cxtmenu` vs. `contextmenuData`/`contextmenu`?

I'm thinking that using the full word `context` may make the API easier to read. 

What are your thoughts?

@mj3cheun @maxkfranz

**Comment by mtwichan** [comment]
Everything looks good to me! We will not be fixing the packaging error in this PR as discussed with @alexcjohnson.

Good to merge after considering the above comment about the naming conventions. Once this PR is merged, I will create another PR to add the example Dash application I created.

💃

**Comment by alexcjohnson** [comment]
> What was the reasoning for naming the context menu property `cxtmenuData`/`cxtmenu` vs. `contextmenuData`/`contextmenu`?
>
> I'm thinking that using the full word `context` may make the API easier to read.

Good call - also to avoid confusion, I've more often seen context abbreviated to `ctx`, so if we use `cxt` I'd be mis-typing it constantly. So yeah, let's use the full word.

**Comment by mj3cheun** [comment]
All comments addressed

**Pull request merged**
